app/decorators/account_decorator.rb

```diff
@@ -3,7 +3,7 @@
 class AccountDecorator < ApplicationDecorator
   delegate :display_name, :email, to: :admin_user, prefix: true
 
-  private
+  protected
```

Comment:
Just curious what changed in ruby to require this

Reply:
Not Ruby, but `draper` gem upgrade (which was needed to support the newer Ruby version).
The issue was here: https://github.com/drapergem/draper/blob/v4.0.2/lib/draper/automatic_delegation.rb#L10-L26

In decorators we [delegate all](https://github.com/3scale/porta/blob/cf17e8c8bd51902c1fa56d8fce31522b36541ee1/app/decorators/application_decorator.rb#L4) methods to the decorated object. But this version of the library only does that when there is no private method with the same name on the decorator. In this case, as `admin_user` was a private method on AccountDecorator, draper tried to call `admin_user` on subclass, but it was not there, so it was failing with: 

```
ActionView::Template::Error (super: no superclass method `admin_user' for #<AccountDecorator:0x00007f8784266d18 @object=#<Account id: 11, org_name: "Testing" ...>
Did you mean?  admin_user_email):
```

So, I removed it from `private_methods` and now it works fine. Not sure it's the best solution though.

Reply:
> respond to two methods

Why two methods?

Reply:
If I understood it correctly, there is an `#admin_user` method in `AccountDecorator` and another `#admin_user` method in `Account`, is that correct?

Reply:
Ah, I see.

`#admin_user` on `Account` is public, as it's an association.

`#admin_user` on `AccountDecorator` was `private` (don't know the reason), I made it `protected`, but I think it shouldn't cause harm if made public too :shrug: 

If we rename the decorator's method, for example,

```
  def admin
    @admin ||= (admin_user || User.new).decorate
  end
```

it doesn't change anything - we still need to make this method public/protected to make it work.

Reply:
It's OK to make it `protected`, but I would rename it anyway, I find it confusing. For instance, if I call `AccountDecorator#admin_user` from inside `AccountDecorator`, which one is called?

If we `delegate_all`, I think it's good idea to not define methods that have the same name as methods in the delegated class.

Reply:
Why is it confusing? Could you elaborate? What name do you suggest? As AccountDecorator is a separate class, it looks alright that it may have such a method that is same name as another class' method, no?

Reply:
> Why is it confusing?

Because `AccountDecorator` had two ways to respond to `admin_user`, isn't that confusing?

> What name do you suggest?

Whatever name, it's was private method, so it didn't matter.

> As AccountDecorator is a separate class, it looks alright that it may have such a method that is same name as another class' method, no?

No if `AccountDecorator` is delegating all undefined methods to that other class, in this case `Account`.

Because of the delegation, `AccountDecorator` responds to all methods defined in `Account`, so if we are defining private methods for `AccountDecorator` itself, better give them different names to not clash. If I call `AccountDecorator#admin_user` from inside `AccountDecorator`, which one is called? The private method in `AccountDecorator` or the public method in `Account`?

Anyway, after the last change, now `AccountDecorator#admin_user` is public method, so it's overriding the one in `Account`. That's fine.

Reply:
> Because `AccountDecorator` had two ways to respond to `admin_user`, isn't that confusing?

I don't think it has 2 ways. The method defined in the `AccountDecorator` is the one that will be called when calling `admin_user`, then it's up the the implementation of this method where to send it further.

I think it's actually a common approach - decorator "decorates" the wrapped object (`Account` in this case) by adding some new methods or overriding others, but the decorator also responds to the methods of the wrapped object.
So, I think having the same method name is fine.

> Whatever name, it's was private method, so it didn't matter.

Yeah, but it was failing exactly because it was private, so I made it non-private, and in fact, I have now also made it public, as I think it actually makes more sense.

> If I call `AccountDecorator#admin_user` from inside `AccountDecorator`, which one is called? The private method in `AccountDecorator` or the public method in `Account`?

As I said above (because I didn't finish reading the comment before replying :grimacing: ) I believe the method (either private or public) in `AccountDecorator` is called always. As with inheritance. First the method inside the same class is checked, and only if not found, superclasses are checked (or wrapped objects in this case).

OK, so we're good now? :smile:

Reply:
Well, it makes no sense going on with the discussion when the confusing situation doesn't exist anymore. But... 😁

> I think it's actually a common approach - decorator "decorates" the wrapped object (Account in this case) by adding some new methods or overriding others, but the decorator also responds to the methods of the wrapped object.

Was a private `#admin_user` overriding the public `#admin_user` from `Account`?. Two comments here:

1. Can a private method override a public method in Ruby?, is that legal? Then the class would implement a different behavior for the same method depending of it being called from inside or outside. That's what I meant by "two ways to respond"
2. This is no real inheritance because `AccountDecorator` is not a subclass of `Account`. The behavior is defined by `drapper`, so it was even more confusing. Apparently, they decided to not allow overriding public methods with private methods, and IMO they did right because that forced us to define our methods as public, which I think is the right approach.

> Yeah, but it was failing exactly because it was private, so I made it non-private, and in fact, I have now also made it public, as I think it actually makes more sense.

Agreed

> OK, so we're good now? 😄

I'll let this go because it's Friday... 😎

Reply:
OK, I see, your point was not about only having the same method name, but rather having the same method name, **BUT** different visibility.

Yeah, I see your point. Indeed, it was a bit confusing, but now it's fixed :)

Reply:
Just for completeness, you can't have two methods of same name in ruby with different visibility. You have just one. Private or not, it overrides the one in superclass. But then you can mark it private or protected or public for that matter in the sub-class. Not a good design though.

In this case this is not even a superclass but some kind of a wrapper. And obviously they decided it was a bad idea there too. Which likely so.

Reply:
Thanks for the explanation @akostadinov